### ISSUE TITLE
why peeked BufferSource.request()  blocked all the time

### ISSUE THREAD

**Issue by LightSun** [square/okio#594]
Hi , All ! You are write a very good framework of IO.
I write a program which used okio to read and write message.
but when i peek a buffer source (origin from socket).  the request method will blocked all the time.
(the client write data and flush ).
if use timeout. it will be closed by timeout.
```java
 BufferedSource peek = in.peek();
                if(!peek.request(8)){   // blocked here all the time.
                    System.out.println("message not all reached.");
                    continue;
      }
```
and if i use this . it will be ok. so . it think this is a bug.
```java
               in.skip(4);
                int len = in.readInt();
                if(!in.request(len)){
                    System.out.println("data not reached.");
                    break;
                }
                byte[] bytes = in.readByteArray(len);
                System.out.println("message reached:  " + new String(bytes, StandardCharsets.UTF_8));
```

the full sample code is here.
```java
public class Server3 {
    // initialize socket and input stream
    private Socket socket = null;
    private ServerSocket server = null;
    private BufferedSource in = null;
    private BufferedSink out = null;

    // constructor with port
    public Server3(int port) {
        //regist

        // starts server and waits for a connection
        try {
            server = new ServerSocket(port);
            System.out.println("Server started");

            System.out.println("Waiting for a client ...");

            socket = server.accept();
            System.out.println("Client accepted");

            // takes input from the client socket
            //in = Okio.buffer(Okio.source(socket.getInputStream()));
            out = Okio.buffer(Okio.sink(socket));
            in = Okio.buffer(Okio.source(socket));

            String line = "";

            // reads message from client until "Over" is sent
            while (!line.equals("Over")) {
                BufferedSource peek = in.peek();
                if(!peek.request(8)){ //blocked all the time
                    System.out.println("message not all reached.");
                    continue;
                }
                peek.skip(4);
                int len = peek.readInt();
                if(!peek.request(len)){
                    System.out.println("message len = " + len + " not all reached.");
                    continue;
                }
                in.skip(8);
                byte[] bytes = in.readByteArray(len);
                System.out.println("message reached:  " + new String(bytes, StandardCharsets.UTF_8));
            }
         //this is ok
        /* while (!line.equals("Over")) {
                in.skip(4);
                int len = in.readInt();
                if(!in.request(len)){
                    System.out.println("data not reached.");
                    break;
                }
                byte[] bytes = in.readByteArray(len);
                System.out.println("message reached:  " + new String(bytes, StandardCharsets.UTF_8));
            } */
            //below is ok.
           /* while (!line.equals("Over")) {
                in.skip(4);
                int len = in.readInt();
                byte[] bytes = in.readByteArray(len);
                System.out.println("message reached:  " + new String(bytes, StandardCharsets.UTF_8));
            }*/
            System.out.println("Closing connection");

            // close connection
            socket.close();
            in.close();
        } catch (IOException i) {
            System.out.println(i);
        }
    }

    public static void main(String args[]) {
        Server3 server = new Server3(5000);
    }
}
```
```java
public class Client3 {
    // initialize socket and input output streams
    private Socket socket = null;
    private BufferedSource input = null;
    private BufferedSink out = null;

    // constructor to put ip address and port
    public Client3(String address, int port) {
        // establish a connection
        try {
            socket = new Socket(address, port);
            System.out.println("Connected");

            // takes input from terminal
            input = Okio.buffer(Okio.source(System.in));

            // sends output to the socket
            out = Okio.buffer(Okio.sink(socket.getOutputStream()));
        } catch (UnknownHostException u) {
            System.out.println(u);
        } catch (IOException i) {
            System.out.println(i);
        }

        // string to read message from input
        String line = "";

        // keep reading until "Over" is input
        while (!line.equals("Over")) {
            try {
                String msg = "sdfdsfjdsfssssssssssssssss";
                out.writeInt(43);
                out.writeInt(msg.length());
                out.write(msg.getBytes(StandardCharsets.UTF_8));
                out.flush();
            } catch (IOException e) {
                e.printStackTrace();
            }
            System.out.println("client write message success,");
            synchronized (this){
                try {
                    this.wait();
                } catch (InterruptedException e) {
                }
            }
        }

        // close the connection
        try {
            input.close();
            out.close();
            socket.close();
        } catch (IOException i) {
            System.out.println(i);
        }
    }

    public static void main(String args[]) {
        Client3 client = new Client3("127.0.0.1", 5000);
    }
}
```

okio version:  2.2.2

**Comment by swankjesse** [comment]
There’s a bug in the peeked streams where they attempt to read upstream before they strictly need to. Will fix.

**Comment by swankjesse** [comment]
This motivates a backport to 1.x. Also releases for both 1.x and 2.x.

**Comment by swankjesse** [comment]
Thanks for the helpful bug report! Made finding and fixing this very easy.

**Comment by LightSun** [comment]
ok. thanks.

**Comment by swankjesse** [comment]
Released as 1.17.4. Fix on the 2.x branch is forthcoming.
